### PR TITLE
TEST/IODEMO: do not use deprecated APIs

### PR DESCRIPTION
**Pull request by changchengx**
## What
Do not use deprecated APIs in io_demo.

### CODE REVIEW

File: test/apps/iodemo/ucx_wrapper.cc [hunk 1]
```diff
@@ -529,11 +529,19 @@ UcxContext::wait_completion(ucs_status_ptr_t status_ptr, const char *title,
...
 void UcxContext::recv_io_message()
 {
-    ucs_status_ptr_t status_ptr = ucp_tag_recv_nb(_worker, &_iomsg_buffer[0],
-                                                  _iomsg_buffer.size(),
-                                                  ucp_dt_make_contig(1),
-                                                  IOMSG_TAG, IOMSG_TAG,
-                                                  iomsg_recv_callback);
+    ucp_request_param_t param;
+
+    param.op_attr_mask = UCP_OP_ATTR_FIELD_DATATYPE |
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:534]: no need to pass datatype for ucp_dt_make_contig(1)

File: test/apps/iodemo/ucx_wrapper.cc [hunk 2]
```diff
@@ -1131,8 +1146,11 @@ void UcxConnection::connect_tag(UcxCallback *callback)
     }
 
     // send local connection id
-    void *sreq = ucp_stream_send_nb(_ep, &_conn_id, 1, dt_int,
-                                    common_request_callback, 0);
+    param.cb.send = common_request_callback_nbx;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:1149]: can we remove the existing common_request_callback, and rename common_request_callback_nbx to common_request_callback?

**Comment by changchengx** [comment]
I'll deal with the comments after June-09.

File: test/apps/iodemo/ucx_wrapper.cc [hunk 3]
```diff
@@ -529,11 +529,17 @@ UcxContext::wait_completion(ucs_status_ptr_t status_ptr, const char *title,
 
 void UcxContext::recv_io_message()
 {
-    ucs_status_ptr_t status_ptr = ucp_tag_recv_nb(_worker, &_iomsg_buffer[0],
-                                                  _iomsg_buffer.size(),
-                                                  ucp_dt_make_contig(1),
-                                                  IOMSG_TAG, IOMSG_TAG,
-                                                  iomsg_recv_callback);
+    ucp_request_param_t param;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:532]: can we make param a global var (static class member) so it will not consume stack space?

Comment (changchengx) [reply test/apps/iodemo/ucx_wrapper.cc:532]: Sounds good. I'll do it at next weekend.

File: test/apps/iodemo/ucx_wrapper.cc [hunk 4]
```diff
@@ -1243,7 +1242,7 @@ bool UcxConnection::send_common(const void *buffer, size_t length,
 
     assert(_ucx_status == UCS_OK);
 
-    ucp_request_param_t param;
+    ucp_request_param_t &param = _context.param;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:1245]: we should make a global param var only for params that could be initialized statically (once) and not changed later. and make a different param var for each case.
This way we can pre-initialize param struct and then just use it during send/recv.
It's not possible when some param fields depend on send function parameters - for example, when passing memh.
Also, it's not needed for slow-path case such as stream operation for exchanging conn_id.

Comment (changchengx) [reply test/apps/iodemo/ucx_wrapper.cc:1245]: @yosefe Could we define class type and constructor to initialize ucp_request_param_t in c++ program? I find it must initialize all the structure members one by one explicitly under ```__cplusplus 199711L```. For example, it needs to initialize all the members until ```.cb.recv``` if we only want to initialize ```op_attr_mask``` and ```cb.recv``` explicitly.

Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.cc:1245]: @changchengx can make these structs static member variables and initialize them like global variables.

Comment (changchengx) [reply test/apps/iodemo/ucx_wrapper.cc]: 1. For slow path(such as ```connect_tag``` which uses ```stream recv/send```), keep the original code there
2. For the ```ucp_request_param_t type object``` that depends on function parameter(such as ```memh```), keep the original code there.
3. Define ```static ucp_request_param_t UcxContext::param``` object and initialize the object in ```UcxContext::init``` function. Then, use this object in ```UcxContext::recv_io_message``` function without constructing it again.

Comment (changchengx) [reply test/apps/iodemo/ucx_wrapper.cc]: @yosefe Passed all CI test cases.

File: test/apps/iodemo/ucx_wrapper.cc [hunk 5]
```diff
@@ -136,6 +136,7 @@ UcxContext::~UcxContext()
     }
 }
 
+ucp_request_param_t UcxContext::param;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:139]: 1. pls rename it to recv_param
2. space line after
3. move it to just before line 90

Comment (changchengx) [reply test/apps/iodemo/ucx_wrapper.cc]: @yosefe Done. Waiting for CI result.

File: test/apps/iodemo/ucx_wrapper.cc [hunk 6]
```diff
@@ -529,11 +535,11 @@ UcxContext::wait_completion(ucs_status_ptr_t status_ptr, const char *title,
...
-    ucs_status_ptr_t status_ptr = ucp_tag_recv_nb(_worker, &_iomsg_buffer[0],
-                                                  _iomsg_buffer.size(),
-                                                  ucp_dt_make_contig(1),
-                                                  IOMSG_TAG, IOMSG_TAG,
-                                                  iomsg_recv_callback);
+    ucs_status_ptr_t status_ptr = ucp_tag_recv_nbx(_worker, &_iomsg_buffer[0],
+                                                   _iomsg_buffer.size(),
+                                                   IOMSG_TAG, IOMSG_TAG,
+                                                   &recv_param);
+
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:542]: remove space line

Comment (changchengx) [reply test/apps/iodemo/ucx_wrapper.cc]: Done

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@changchengx pls squash

**Comment by changchengx** [comment]
@yosefe squashed

**Review by yosefe — APPROVED** [review]

**Pull request merged**